src/main/java/com/powsybl/caseserver/service/S3CaseService.java

```diff
@@ -0,0 +1,574 @@
+        } catch (IOException e) {
+            throw CaseException.createTempDirectory(caseUuid, e);
+        }
+        try {
+            tempCasePath = tempdirPath.resolve(filename);
+            try {
+                contentInitializer.accept(tempCasePath);
+            } catch (Exception e) {
+                throw CaseException.createUInitTempFileError(caseUuid, e);
+            }
+            // after this line, need to cleanup the file
+            try {
+                try {
+                    return f.apply(tempCasePath);
+                } catch (Exception e) {
+                    throw CaseException.createFileNotImportable(tempdirPath);
+                }
+            } finally {
+                try {
+                    Files.delete(tempCasePath);
+                } catch (IOException e) {
+                    LOGGER.error("Error cleaning up temporary case file", e);
+                }
+            }
+        } finally {
+            try {
+                if (Files.exists(tempdirPath)) {
+                    FileUtils.deleteDirectory(tempdirPath.toFile());
+                }
+            } catch (IOException e) {
+                LOGGER.error("Error cleaning up temporary case directory: " + tempdirPath, e);
+            }
+        }
+    }
+
+    // downloads from s3 and cleanup
+    public <R, T extends Exception> R withS3DownloadedTempPath(UUID caseUuid, FailableFunction<Path, R, T> f) {
+        return withS3DownloadedTempPath(caseUuid, null, f);
+    }
+
+    public <R, T extends Exception> R withS3DownloadedTempPath(UUID caseUuid, String caseFileKey, FailableFunction<Path, R, T> f) {
+        String nonNullCaseFileKey = Objects.requireNonNullElse(caseFileKey, uuidToKeyWithOriginalFileName(caseUuid));
+        String filename = parseFilenameFromKey(nonNullCaseFileKey);
+        return withTempCopy(caseUuid, filename, path ->
+                        s3Client.getObject(GetObjectRequest.builder().bucket(bucketName).key(nonNullCaseFileKey).build(), path), f);
+    }
+
+    @Override
+    public String getFormat(UUID caseUuid) {
+        return getCaseMetaDataEntity(caseUuid).getFormat();
+    }
+
+    public String getCompressionFormat(UUID caseUuid) {
+        return getCaseMetaDataEntity(caseUuid).getCompressionFormat();
+    }
+
+    public String getOriginalFilename(UUID caseUuid) {
+        return getCaseMetaDataEntity(caseUuid).getOriginalFilename();
+    }
+
+    // key format is "gsi-cases/UUID/filename"
+    private UUID parseUuidFromKey(String key) {
+        int firstSlash = key.indexOf(DELIMITER);
+        int secondSlash = key.indexOf(DELIMITER, firstSlash + 1);
+        return UUID.fromString(key.substring(firstSlash + 1, secondSlash));
+    }
+
+    private String parseFilenameFromKey(String key) {
+        int firstSlash = key.indexOf(DELIMITER);
+        int secondSlash = key.indexOf(DELIMITER, firstSlash + 1);
+        return key.substring(secondSlash + 1);
+    }
```

Comment:
Both methods parseUuidFromKey and parseFilenameFromKey are finding the indices of firstSlash and secondSlash, which are the same calculations in both. We can calculate these indices only once and reuse them

Reply:
not sure if it's worth it. If we wanted to do this, we should create a pojo with all the parsed info and then someone can read the differents parts..